### PR TITLE
Type storage in enact.

### PR DESCRIPTION
**Pull request by lcrh**
This adds type storage to enact, which is a step towards representing dynamic types.

### CODE REVIEW

File: src/enact/types.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# Copyright 2023 Agentic.AI Corporation.
```
Comment (nathanmartz) [inline src/enact/types.py:1]: 2024

File: src/enact/types.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                         f'{json_value}')
+      return ResourceType(TypeKey.from_dict(value))
+    raise ValueError(f'Unknown type descriptor: {json_value}')
+
+@dataclasses.dataclass
+class Int(TypeDescriptor):
+  """Describes an int value."""
+
+  def to_json(self):
+    return INT_NAME
```
Comment (nathanmartz) [inline src/enact/types.py:121]: Recommend moving these names into the type dataclasses which would make the logic a bit more self contained. So you could do:

@dataclasses.dataclass
class Int(TypeDescriptor):
  """Describes an int value."""
  NAME: ClassVar[str] = 'int

  def to_json(self):
    return Int.NAME

File: src/enact/type_inference.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  if t == bytes:
+    return types.Bytes()
+  if t == list:
+    return types.List()
+  if t == dict:
+    return types.Dict()
+  if hasattr(t, '__origin__') and t.__origin__ in (types.List, list):
+    return types.List()
+  if hasattr(t, '__origin__') and t.__origin__ in (types.Dict, dict):
+    return types.Dict()
```
Comment (nathanmartz) [inline src/enact/type_inference.py:40]: This long series of conditions is very similar to the logic in types.py. Makes me wonder if you could just have something like:

```
BASIC_TYPES = {
  int: Int
  bool: Bool
}
```

Then you could just look it up here, iterate through it in types.py, etc.

Comment (nathanmartz) [reply src/enact/type_inference.py:40]: Note that the iteration would depend on my other suggestion of moving NAME into the individual classes.

File: src/enact/interfaces.py [hunk 4]
```diff
@@ -16,24 +16,20 @@
...
-Json = Union[JsonLeaf, List['Json'], Dict[str, 'Json']]
-
 
 PRIMITIVES = (int, float, str, bytes, bool, type(None))
 
 
 Primitives = Union[
-  int, float, str, bytes, bool,
-  None]
+  int, float, str, bytes, bool, None]
```
Comment (nathanmartz) [inline src/enact/interfaces.py:32]: Feels like PRIMITIVES and Primitives should go in types.py now that it exists.

**Review by nathanmartz — APPROVED** [review]
Some suggestions for simplification, but I generally like how types can be added to the store along with instances of that type.

File: src/enact/types.py (hunk 2 above)
Comment (lcrh) [reply src/enact/types.py]: Done

File: src/enact/type_inference.py (hunk 3 above)
Comment (lcrh) [reply src/enact/type_inference.py:40]: Done

File: src/enact/interfaces.py (hunk 4 above)
Comment (lcrh) [reply src/enact/interfaces.py]: Done

**Pull request merged**